crates/futures/src/polyfill.rs

```diff
@@ -0,0 +1,208 @@
+    Reflect::get(array, &JsValue::from(index))
+        .expect(&format!("Array should contain the index {}", index))
+}
+
+// Atomics.waitAsync always returns a promise. Throws standard errors
+// for parameter validation.  The promise is resolved with a string as from
+// Atomics.wait, or, in the case something went completely wrong, it is
+// rejected with an error string.
+pub fn wait_async_with_timeout(index: u32, value: i32, timeout: f64) -> Result<Promise, JsValue> {
+    let memory_buffer = wasm_bindgen::memory()
+        .dyn_into::<WebAssembly::Memory>()
+        .expect("Should cast a memory to WebAssembly::Memory")
+        .buffer();
+
+    let indexed_array = Int32Array::new(&memory_buffer);
+
+    if !indexed_array.buffer().has_type::<SharedArrayBuffer>() {
+        console_log!("polyfill, not a SharedArrayBuffer");
+        return Err(Error::new("Indexed array must be created from SharedArrayBuffer").into());
+    }
+
+    // Optimization, avoid the helper thread in this common case.
+    if Atomics::load(&indexed_array, index)? != value {
+        console_log!("polyfill, not-equal");
+        return Ok(Promise::resolve(&JsString::from("not-equal")));
+    }
+
+    // General case, we must wait.
+
+    console_log!("polyfill, general case");
+
+    Ok(Promise::new(
+        &mut move |resolve: Function, reject: Function| {
+            let helper = alloc_helper();
+            let helper_ref = helper.clone();
+
+            let onmessage_callback = Closure::once_into_js(Box::new(move |e: MessageEvent| {
+                // Free the helper early so that it can be reused if the resolution
+                // needs a helper.
+                free_helper(&helper_ref);
+                match String::from(
+                    get_array_item(&e.data(), 0)
+                        .as_string()
+                        .expect("data[0] should return a String"),
+                )
+                .as_str()
+                {
+                    "ok" => {
+                        resolve
+                            .call1(&JsValue::NULL, &get_array_item(&e.data(), 1))
+                            .expect("Should successfully call a resolve callback");
+                    }
+                    "error" => {
+                        // Note, rejection is not in the spec, it is an artifact of the polyfill.
+                        // The helper already printed an error to the console.
+                        reject
+                            .call1(&JsValue::NULL, &get_array_item(&e.data(), 1))
+                            .expect("Should successfully call a reject callback");
+                    }
+                    // it's not specified in the proposal yet
+                    _ => (),
+                }
+            })
+                as Box<dyn FnMut(MessageEvent)>);
+            helper.set_onmessage(Some(onmessage_callback.as_ref().unchecked_ref()));
+
+            // onmessage_callback.forget();
```

Comment:
The instance of `Closure` here needs to be persisted with the `Worker` to ensure it's not deallocated on the Rust side of things, but this is basically stored with `helper_ref` I think

Reply:
Do you mean to create something like
```rust
struct WorkerData {
  worker: Worker,
  closure: Closure,
}
```
?